src/commands/signzone.rs

```diff
@@ -499,20 +545,97 @@ impl SignZone {
+            let n_records = records.len();
+            let pb = if matches!(&log, Some(log) if log.is_terminal() && self.progress) {
+                let pb = ProgressBar::new(0).with_message("Signing");
+                pb.set_draw_target(ProgressDrawTarget::stderr_with_hz(1));
+                pb.set_style(
+                    ProgressStyle::with_template(PROGRESS_STYLE)
+                        .unwrap()
+                        .with_key("eta", |state: &ProgressState, w: &mut dyn Write| {
+                            write!(w, "{:.1}s", state.eta().as_secs_f64()).unwrap()
+                        })
+                        .progress_chars("#>-"),
+                );
+                Some(pb)
+            } else {
+                None
+            };
             let extra_records = records
                 .sign(
                     &apex,
                     self.expiration,
                     self.inception,
                     keys.as_slice(),
                     !self.do_not_add_keys_to_zone,
+                    |inc_pos, inc_len, new_phase| {
+                        if let Some(pb) = &pb {
+                            if inc_len > 0 {
+                                pb.inc_length(inc_len as u64);
+                            }
+                            if inc_pos > 0 {
+                                pb.inc(inc_pos as u64);
+                            }
+                            if let Some(new_phase) = new_phase {
+                                pb.set_message(new_phase);
+                            }
+                        }
+                    },
                 )
                 .unwrap();
             records.extend(extra_records.into_iter().map(Record::from_record));
+
+            if let Some(pb) = pb {
+                let len = pb.length().unwrap();
+                let elapsed = pb.elapsed();
+                pb.finish_and_clear();
+                if self.verbose {
+                    if let Some(log) = &mut log {
+                        writeln!(
+                            log,
+                            "Signed {n_records} records in {len} steps in {}",
+                            HumanDuration(elapsed)
+                        );
+                    }
+                }
+            }
         }
 
+        let n_records = records.len();
+        let pb = if matches!(&log, Some(log) if log.is_terminal() && self.progress) {
+            let num_families = records.families().count();
+            let pb = ProgressBar::new(num_families as u64);
+            pb.set_draw_target(ProgressDrawTarget::stderr_with_hz(1));
+            pb.set_style(
+                ProgressStyle::with_template(PROGRESS_STYLE)
+                    .unwrap()
+                    .with_key("eta", |state: &ProgressState, w: &mut dyn Write| {
+                        write!(w, "{:.1}s", state.eta().as_secs_f64()).unwrap()
+                    })
+                    .progress_chars("#>-"),
+            );
```

Comment:
This style should probably be saved somewhere so we can reuse it.

Reply:
Yup, there's far too much repetitive boilerplate here, but not being familiar with the progress bar crate I wanted to first get something working.